From ab0b8743196c3dc1592537aa1ff188e5001fb42a Mon Sep 17 00:00:00 2001
From: Darryl James <dljames@chromium.org>
Date: Thu, 22 Jan 2026 10:57:07 -0800
Subject: [PATCH] Reland "[Vertical Tabs] Fix TabStripModelObserver Resetting Multiple Times"

This reverts commit 7ba5cd1c17dc24ed54e05864421010445d57bd30.

Reason for revert: Unlikely this CL was causing the test failures as no VerticalTabStrip specific code modifications are included in this patch.

Original change's description:
> Revert "[Vertical Tabs] Fix TabStripModelObserver Resetting Multiple Times"
>
> This reverts commit 9a2095a6bd525987915d98a90f654560f5e7bb74.
>
> Reason for revert: breaks multiple Linux and Mac bots 
> First failure: https://ci.chromium.org/ui/p/chromium/builders/ci/linux-bfcache-rel/80596 
>
> Original change's description:
> > [Vertical Tabs] Fix TabStripModelObserver Resetting Multiple Times
> >
> > In TabStripModel, the 'tab_strip_ui_was_set_' flag was not being reset
> > when the observer was removed. This could lead to crashes if a new
> > observer was set during runtime.
> >
> > Unit tests have been added to verify this fix.
> >
> > Change-Id: Id402a81a2cc6f7f90beccaa884cd1c2e2ab9f94f
> > Bug: 468100896
> > Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/7507604
> > Commit-Queue: Darryl James <dljames@chromium.org>
> > Reviewed-by: Eshwar Stalin <estalin@chromium.org>
> > Cr-Commit-Position: refs/heads/main@{#1572766}
>
> Bug: 468100896
> No-Presubmit: true
> No-Tree-Checks: true
> No-Try: true
> Change-Id: I2f51dde1bf6ff8617d89ef50d00ac1950c57c82a
> Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/7510603
> Commit-Queue: Rubber Stamper <rubber-stamper@appspot.gserviceaccount.com>
> Bot-Commit: Rubber Stamper <rubber-stamper@appspot.gserviceaccount.com>
> Auto-Submit: Viktor Semeniuk <vsemeniuk@google.com>
> Owners-Override: Viktor Semeniuk <vsemeniuk@google.com>
> Cr-Commit-Position: refs/heads/main@{#1572822}

Bug: 468100896
Change-Id: I00929fb3c572eb898ad6ad8e5bd2874d34d42da9
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/7511472
Commit-Queue: Darryl James <dljames@chromium.org>
Reviewed-by: Eshwar Stalin <estalin@chromium.org>
Bot-Commit: Rubber Stamper <rubber-stamper@appspot.gserviceaccount.com>
Cr-Commit-Position: refs/heads/main@{#1573180}
---

--- a/chrome/browser/ui/tabs/tab_strip_model.cc
+++ b/chrome/browser/ui/tabs/tab_strip_model.cc
@@ -342,6 +342,14 @@ void TabStripModel::AddObserver(TabStrip
 
 void TabStripModel::RemoveObserver(TabStripModelObserver* observer) {
   observer->StoppedObserving(TabStripModelObserver::ModelPasskey(), this);
+
+  // Reset `tab_strip_ui_was_set_` flag if it is the first observer in the list.
+  // The `SetTabStripUI` function ensures it is the first observer added.
+  if (tab_strip_ui_was_set_ && !observers_.empty() &&
+      &*observers_.begin() == observer) {
+    tab_strip_ui_was_set_ = false;
+  }
+
   observers_.RemoveObserver(observer);
 }
 
